packages/utils/src/libp2p/dialHelper.ts

```diff
@@ -44,121 +45,189 @@ export type DialResponse =
     }
   | {
       status: DialStatus.DIAL_ERROR
-      error: string
       dhtContacted: boolean
     }
   | {
       status: DialStatus.DHT_ERROR
-      error: Error
       query: PeerId
     }
 
-function renderPeerStoreAddresses(addresses: Address[], delimiter: string = '\n  '): string {
-  return addresses.map((addr: Address) => addr.multiaddr.toString()).join(delimiter)
+// Make sure that Typescript fails to build tests if libp2p API changes
+type ReducedPeerStore = { peerStore: Pick<LibP2P['peerStore'], 'get'> }
+type ReducedDHT = { peerRouting: Pick<LibP2P['peerRouting'], '_routers' | 'findPeer'> }
+type ReducedLibp2p = ReducedDHT & ReducedPeerStore & Pick<LibP2P, 'dialProtocol'>
+
+function printPeerStoreAddresses(msg: string, addresses: Address[], delimiter: string = '\n  '): string {
+  return msg.concat(addresses.map((addr: Address) => addr.multiaddr.toString()).join(delimiter))
 }
 
-async function doDial(
-  libp2p: LibP2P,
+/**
+ * Performs a dial attempt and handles possible errors.
+ * @param libp2p Libp2p instance
+ * @param destination which peer to dial
+ * @param protocol which protocol to use
+ * @param opts timeout options
+ */
+async function attemptDial(
+  libp2p: Pick<LibP2P, 'dialProtocol'>,
   destination: PeerId,
   protocol: string,
-  opts: {
-    timeout: number
-    signal: AbortSignal
-  }
-): Promise<DialResponse> {
-  let err: any
+  opts: Required<TimeoutOpts>
+): Promise<
+  | DialResponse
+  | {
+      status: InternalDialStatus.CONTINUE
+    }
+> {
   let struct: Awaited<ReturnType<LibP2P['dialProtocol']>> | null
 
-  let addresses = (libp2p.peerStore.get(destination)?.addresses ?? []).map((addr) => addr.multiaddr.toString())
-
-  // Try to use known addresses
-  if (addresses.length > 0) {
-    try {
-      struct = await libp2p.dialProtocol(destination, protocol, { signal: opts.signal })
-    } catch (_err) {
-      err = _err
+  try {
+    struct = await libp2p.dialProtocol(destination, protocol, { signal: opts.signal })
+  } catch (err) {
+    logError(`Error while dialing ${destination.toB58String()} directly.`)
+    if (err?.message) {
+      logError(`Dial error: ${err.message}`)
     }
   }
 
-  if (struct != null) {
+  // Libp2p's return types tend to change every now and then
+  if (struct != null && struct != undefined) {
```

Comment:
This is the same as `struct != null` isn't it?

Reply:
changed